File: ecosystem/epistemic/eigenius/orchestration/src/observability/guard.ts
// Copyright 2026 The Eigenius Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Per-RPC observability guard for the orchestrator. Mirror of
 * `kernel/src/observability/guard.rs`'s `RpcGuard`.
 *
 * Wrap an async handler body with `withRpcGuard` so the entry / exit
 * events fire automatically for every return path including thrown
 * errors:
 *
 * ```ts
 * await withRpcGuard(operation.COMPONENT_DISPATCH, async (mark) => {
 *   try {
 *     return await dispatch(...);
 *   } catch (e) {
 *     mark.fail("dispatch_failed");
 *     throw e;
 *   }
 * });
 * ```
 *
 * The guard emits `request received` at debug on entry and a
 * completion event at debug on success (with `latency_ms`), or at
 * `warn` if the handler called `mark.fail("kind")`.
 */

import { debug, warn } from "./log.ts";

export interface FailMark {
  /** Mark the request as failed with the given stable error kind. */
  fail(kind: string): void;
}

export async function withRpcGuard<T>(
  operation: string,
  body: (mark: FailMark) => Promise<T>,
): Promise<T> {
  const started = performance.now();
  let errorKind: string | undefined;
  const mark: FailMark = {
    fail(kind: string) {
      errorKind = kind;
    },
  };

  debug(operation, "request received");

  try {
    const result = await body(mark);
    return result;
  } catch (e) {
    if (errorKind === undefined) {
      // Unhandled throw — record it as a generic failure so the
      // completion event still goes out at warn level.
      errorKind = "unhandled_exception";
    }
    throw e;
  } finally {
    const latencyMs = Math.round(performance.now() - started);
    if (errorKind !== undefined) {
      warn(operation, "request failed", {
        latency_ms: latencyMs,
        error_kind: errorKind,
      });
    } else {
      debug(operation, "request completed", { latency_ms: latencyMs });
    }
  }
}

File: ecosystem/epistemic/eigenius/kernel/src/observability/guard.rs
// Copyright 2026 The Eigenius Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! RAII guard for RPC handler observability.
//!
//! `RpcGuard::start("kernel.rpc.foo")` emits a `debug` event for the
//! request and starts a timer. When the guard drops — at the end of
//! the handler, regardless of success / early-`?` / panic — it emits
//! a completion event carrying `latency_ms`. Optional `fail(kind)`
//! upgrades the completion event to `warn` and adds `error_kind`.
//!
//! This keeps every handler's instrumentation to one line at entry
//! plus (optional) one `guard.fail(...)` on the failure paths,
//! without having to repeat the log call at every return.

use crate::observability::field;

pub struct RpcGuard {
    operation: &'static str,
    started: std::time::Instant,
    error_kind: Option<&'static str>,
}

impl RpcGuard {
    /// Start a guard, emitting the `request received` debug event and
    /// recording the start time.
    pub fn start(operation: &'static str) -> Self {
        tracing::debug!({ field::OPERATION } = operation, "request received");
        Self {
            operation,
            started: std::time::Instant::now(),
            error_kind: None,
        }
    }

    /// Mark the request as failed with the given stable error kind.
    /// On drop, the guard will emit a `warn` completion event instead
    /// of the default `debug`. Last call wins if invoked more than
    /// once.
    pub fn fail(&mut self, kind: &'static str) {
        self.error_kind = Some(kind);
    }
}

impl Drop for RpcGuard {
    fn drop(&mut self) {
        let latency_ms = self.started.elapsed().as_millis() as u64;
        match self.error_kind {
            Some(kind) => tracing::warn!(
                { field::OPERATION } = self.operation,
                { field::LATENCY_MS } = latency_ms,
                { field::ERROR_KIND } = kind,
                "request failed"
            ),
            None => tracing::debug!(
                { field::OPERATION } = self.operation,
                { field::LATENCY_MS } = latency_ms,
                "request completed"
            ),
        }
    }
}
